### PR TITLE
Make `binary_cross_entropy` differentiable wrt `target`

### LINKED ISSUE #56683: binary_cross_entropy misses derivative for target even if the variable is unused for backward
I don't understand why it breaks in the forward pass? The backprop through the loss variable may not even be required (e.g. when it's used as matching affinity for non-differentiated algorithm).

I would porpose:
- it should have a derivative formula for target (it can be useful for bringing two probability distributions together in contrastive formulations)
- if it doesn't have it, it should fail at backprop time when the variable is known to be backpropped through in the graph

```python
import torch
import torch.nn.functional as F

a = torch.ones(1, 2) / 2
a.requires_grad_()
b = torch.ones(1, 2) / 2

matching_matrix_unused_for_backprop = F.binary_cross_entropy(a, b.requires_grad_(), reduce = 'none')

#  File ".../vadim/prefix/miniconda/lib/python3.8/site-packages/torch/nn/functional.py", line 2525, in binary_cross_entropy
#    return torch._C._nn.binary_cross_entropy(
# RuntimeError: the derivative for 'target' is not implemented
```

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @albanD @gqchen @pearu @nikitaved @soulitzer

### PR DESCRIPTION
**Pull request by nikitaved**
As per title. Resolves https://github.com/pytorch/pytorch/issues/56683.
`gradgradcheck` will fail once `target.requires_grad() == True` because of the limitations of the current double backward implementation.

### CODE REVIEW

File: test/test_nn.py [hunk 1]
```diff
@@ -8671,6 +8671,35 @@ def test_pdist_cuda_gradgrad_unimplemented(self):
...
+            gradgradcheck(F.binary_cross_entropy, [input, target])
+
+            # now with diffentiable target
+            target.requires_grad_(True)
+            # grad wrt target is not a native function (in a sense that it is
+            # implemented on top of "high-level" functions, and does not have
+            # explicit device-specific dispatches)
+            # and uses in-place operations that trigger vmap failures in gradcheck,
+            # hence the flag `check_batched_grad=False`
+            # TODO: use native kernels to circumvent this limitation.
```
Comment (albanD) [inline test/test_nn.py:8690]: You don't need to do that, you can actually do something like binary_cross_entropy_double_backward to make it work without a native function.

File: test/test_nn.py [hunk 2]
```diff
@@ -8671,6 +8671,35 @@ def test_pdist_cuda_gradgrad_unimplemented(self):
...
+            gradcheck(F.binary_cross_entropy, [input, target], check_batched_grad=False)
+            with self.assertRaisesRegex(RuntimeError, "not implemented"):
+                # binary_cross_entropy_backward only returns grads wrt input,
```
Comment (albanD) [inline test/test_nn.py:8693]: I'm not sure to understand what this comment is trying to explain.
Is it just: "we don't support double backward wrt to target" ?

We'll only need to update `binary_cross_entropy_target_backward` to be differentiable (by removing inplace conditioned on the current GradMode) and add a formula for target: for `binary_cross_entropy_backward`

**Review by albanD — COMMENTED** [review]
Thanks for the update!
The change looks good to me.

I'm not sure if all the comments are necessary though, we don't need to justify why we don't implement a give formula. Especially because such comment can become stale very quickly.

File: test/test_nn.py (hunk 1 above)
Comment (nikitaved) [reply test/test_nn.py:8690]: Shall I just remove it?

File: test/test_nn.py (hunk 2 above)
Comment (nikitaved) [reply test/test_nn.py:8693]: but `binary_cross_entropy_backward` returns `grad_wrt_input` and does not handle `grad_wrt_target`, would that still work?

File: test/test_nn.py (hunk 1 above)
Comment (albanD) [reply test/test_nn.py:8690]: I would remove the comment here. Yes

File: test/test_nn.py (hunk 2 above)
Comment (albanD) [reply test/test_nn.py:8693]: You can have the backward split into two part: one function that compute grads wrt to input and another that compute grads wrt to target.
But if both of these use target, then the double backward will need the double backward of both to compute the double gradient for target.

That will work yes, if you think of this as a single backward, it's the same as if you're doing `res = f1(inp) + f2(inp)` and ask for gradients wrt `inp`.

File: tools/autograd/derivatives.yaml [hunk 3]
```diff
@@ -1361,9 +1361,13 @@
 
 - name: binary_cross_entropy(Tensor self, Tensor target, Tensor? weight=None, int reduction=Mean) -> Tensor
   self: binary_cross_entropy_backward(grad, self, target, weight, reduction)
+  target: binary_cross_entropy_target_backward(grad, self, target, weight, reduction)
 
 - name: binary_cross_entropy_backward(Tensor grad_output, Tensor self, Tensor target, Tensor? weight=None, int reduction=Mean) -> Tensor
   self: binary_cross_entropy_double_backward(grad_output, grad, self, target, weight, reduction)
+  # see the comment in the function `test_binary_cross_entropy_grads` from `test_nn.py`,
+  # where it is explained why grads wrt `target` are forbidden
```
Comment (albanD) [inline tools/autograd/derivatives.yaml:1369]: nit this comment is not needed anymore.

**Review by albanD — APPROVED** [review]
Looks good!
Thanks for the update!

File: test/test_nn.py (hunk 2 above)
Comment (nikitaved) [reply test/test_nn.py]: If I decide to split it into two functions, would it be possible to access grad_target and grad_input in both double backwards for input and target?

**Pull request closed without merging**